### PR TITLE
Add SemiFlatMap method to various Monad Transformers

### PR DESCRIPTION
**Pull request by RawToast**
This is based on the handy `semiFlatmap` methods within the transformers used in [Cats](https://github.com/typelevel/cats). For example:

```
def semiflatMap[B](f: A => F[B])(implicit F: Monad[F]): OptionT[F, B] =
    flatMap(a => OptionT.liftF(f(a)))
```

https://github.com/typelevel/cats/blob/a5c44b46dd7b4622d41e7f273bb25246fa1985d0/core/src/main/scala/cats/data/OptionT.scala#L46-L47

### CODE REVIEW

File: src/Relude_OptionT.re [hunk 1]
```diff
@@ -26,6 +26,10 @@ module WithMonad = (M: BsAbstract.Interface.MONAD) => {
     (aToB, OptionT(mOptionA)) =>
       OptionT(M.map(optionA => Relude_Option.map(aToB, optionA), mOptionA));
 
+  let semiFlatMap: 'a 'b. ('a => option('b), t('a)) => t('b) =
```
Comment (andywhite37) [inline src/Relude_OptionT.re]: I was looking at cats to refresh my memory on `semi`/`subflatMap`, and I'm wondering if this signature is actually `subFlatMap`?  I recall using `subflatMap` in the past in cats, and it made sense to me b/c you're flatMapping the inner/sub monad.

https://github.com/typelevel/cats/blob/bfc264eddd621bc81dff681060be37656e04f72b/core/src/main/scala/cats/data/OptionT.scala#L64-L65

I would love to have both `sub` and `semiFlatMap` if you agree and would like to rename this to `subFlatMap`, and add `semiFlatMap`?  Or let me know if you disagree with this - I might be just confused.

File: src/Relude_ReaderT.re [hunk 2]
```diff
@@ -51,6 +51,16 @@ module WithMonad = (M: BsAbstract.Interface.MONAD) => {
               rToMB(r);
             },
           ),
+      )
+
+  let semiFlatMap: 'r 'a 'b. ('a => M.t('b), t('r, 'a)) => t('r, 'b) =
+    (aToMA, ReaderT(rToMA)) =>
```
Comment (andywhite37) [inline src/Relude_ReaderT.re]: This looks like what I'd expect for `semiFlatMap`

File: src/Relude_ResultT.re [hunk 3]
```diff
@@ -37,6 +37,10 @@ module WithMonad = (M: BsAbstract.Interface.MONAD) => {
     (aToB, ResultT(mResultA)) =>
       ResultT(M.map(resultA => Result.map(aToB, resultA), mResultA));
 
+  let semiFlatMap: 'a 'b 'e. ('a => Result.t('b, 'e),  t('a, 'e)) => t('b, 'e) =
```
Comment (andywhite37) [inline src/Relude_ResultT.re]: Same comment as `OptionT` - is this `subFlatMap`?  I'd love to have both `sub` and `semi` here too.

https://github.com/typelevel/cats/blob/c56edd78e6accfc276bfffad15d08d309a76076e/core/src/main/scala/cats/data/EitherT.scala#L364-L365

**Comment by RawToast** [comment]
Sure, I’ll double check this. It’s not hard to get the two mixed up 😸

I might take a quick look at introducing both functions.

**Comment by andywhite37** [comment]
Cool, let me know when you're done with changes, and I can merge it

**Comment by RawToast** [comment]
That's about as far as I can take it for now. Feel free to squash/rebase 😸 As I get more familiar with the library, I'll see if there's anything else I can add.

**Comment by andywhite37** [comment]
Sweet, looks good to me!  Thanks again!

**Pull request merged**